Name: Jeremy Tan <br>
ID: 204947 <br>
Date Accomplished: March 8, 2022

In [2]:
import numpy as np

### Problem A

Let  denote the quality of the n-th item produced by a production system with X<sub>n</sub> = 0  meaning "good" and  X<sub>n</sub> = 1 meaning "defective".  Suppose that  evolves as a Markov chain whose transition probability matrix is given by the following:




In [3]:
A = [[0.99, 0.12], [0.01, 0.88]]
A = np.array(A)
A

array([[0.99, 0.12],
       [0.01, 0.88]])

What is the probability that the 20th item is defective given that the first item is defective? Simulate.

In [4]:
def production(number):
    x = np.array([[0], [1]]) # state distribution vector: the first item is defective
    for i in range(1,number):
        x = np.matmul(A,x) # Matrix Multiplication
    return x
print(production(20)) # until the 20th item

[[0.85759767]
 [0.14240233]]


In [5]:
print(production(20)[1,0]*100)

14.240232649180864


<p style="font-size:16px">There is a <b>14.24%</b> chance the 20th item is defective.</p> 
<p style="font-size:16px">Note: The first item out of twenty was the initial defective item.</p>

### Problem B

A bus in a mass transit system is operating on a continuous route with intermediate stops.  The arrival of the bus at a stop is classified into one of three states, namely early arrival, on-time arrival, and late arrival. 

Suppose that the successive states form a Markov chain with transition probability matrix

In [6]:
A = [[0.5,0.2,0.1], [0.4,0.5,0.2], [0.1,0.3,0.7]]
A = np.array(A)
A

array([[0.5, 0.2, 0.1],
       [0.4, 0.5, 0.2],
       [0.1, 0.3, 0.7]])

Over a long period of time, what fraction of stops can be expected to be late? Simulate.

I'm looking for the convergent transition matrix of A. As the number of iterations, <b>k</b>, approaches infinity, A<sup>k</sup> converges to a value.

Each step in the markov chain is defined by the matrix A multiplied by itself.

In [20]:
def bus(k):
    x = np.matmul(A,A) # state distribution vector
    for i in range(1,k): 
        x = np.matmul(A,x)
    return x
print(bus(100)) # we take 100 iterations to simulate a long period of time

[[0.225 0.225 0.225]
 [0.35  0.35  0.35 ]
 [0.425 0.425 0.425]]


In [32]:
bus(100)[:,[0]] # the third row corresponds to the probability of being late

array([[0.225],
       [0.35 ],
       [0.425]])

<p style="font-size:16px">Over a long period of time, around 42.5% or <b>17 out of every 40</b> of the bus's stops will be late.</p>